File: utils/syscall_wrapper/copy_file_range.c
/*
 * a syscall wrapper in C for copy_file_range(2) test
 * author: Jianhong Yin <yin-jianhong@163.com>
 */

#define _GNU_SOURCE
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/syscall.h>
#include <unistd.h>
#include <string.h>

ssize_t copy_file_range(int fd_in, loff_t *off_in,
			int fd_out, loff_t *off_out,
			size_t len, unsigned int flags)
{
#ifndef __NR_copy_file_range
#define __NR_copy_file_range 326
#endif

	return syscall(__NR_copy_file_range, fd_in, off_in, fd_out,
		       off_out, len, flags);
}

int main(int argc, char **argv)
{
	unsigned len, ret;
	int fd_in;
	int fd_out;
	loff_t off_in = 0;
	loff_t off_out = 0;

	char *dst = NULL;
	char *src = NULL;
	char *dst_file = NULL;
	char *src_file = NULL;
	char *dst_off = NULL;
	char *src_off = NULL;
	struct stat stat;
	char *endptr;

	if (argc < 3) {
		fprintf(stderr, "Usage: %s <destination[:offset]> <<source|->[:offset]> [len]\n", argv[0]);
		fprintf(stderr, "`note: source is '-' means that 'fd_in' equal 'fd_out' # see copy_file_range(2)\n");
		fprintf(stderr, "  e.g1: %s testfile:256  -  64\n", argv[0]);
		fprintf(stderr, "  e.g2: %s fileout  filein  1024\n", argv[0]);
		exit(EXIT_FAILURE);
	}

	dst = strdup(argv[1]);
	src = strdup(argv[2]);

	dst_file = strsep(&dst, ":");
	dst_off = strsep(&dst, ":");
	if (dst_off != NULL)
		off_out = strtoll(dst_off, &endptr, 0);

	src_file = strsep(&src, ":");
	src_off = strsep(&src, ":");
	if (src_off != NULL)
		off_in = strtoll(src_off, &endptr, 0);

	fd_out = open(dst_file, O_RDWR|O_CREAT, 0666);
	if (fd_out == -1) {
		perror("open dst file");
		exit(EXIT_FAILURE);
	}

	if (strcmp("-", src_file) == 0 || src_file[0] == '\0') {
		fd_in = fd_out;
	} else {
		fd_in = open(src_file, O_RDONLY);
		if (fd_in == -1) {
			perror("open src file)");
			exit(EXIT_FAILURE);
		}
	}

	if (fstat(fd_in, &stat) == -1) {
		perror("fstat");
		exit(EXIT_FAILURE);
	}
	len = stat.st_size;

	if (fd_in == fd_out && dst_off == NULL)
		off_out=len;

	if (argv[3] != NULL) {
		len = strtoll(argv[3], &endptr, 0);
	}

	do {
		ret = copy_file_range(fd_in, &off_in, fd_out, &off_out, len, 0);
		if (ret == -1) {
			perror("copy_file_range");
			exit(EXIT_FAILURE);
		}
	        printf("ret=%d, len=%zu\n", ret, len);
		if (ret == 0)
			break;
		len -= ret;
	} while (len > 0);

	close(fd_in);
	if (fd_in != fd_out)
		close(fd_out);
	exit(EXIT_SUCCESS);
}

File: utils/syscall_wrapper/copy_file_range.pl
#!/usr/bin/perl
#
# a syscall wrapper in perl for copy_file_range(2) test
# author: Jianhong Yin <yin-jianhong@163.com>
#
# ref: https://github.com/tcler/linux-network-filesystems/blob/master/utils/h2ph.sh
require 'syscall.ph';  # may need to run h2ph first ^^^
use POSIX ();

my $argc = @ARGV;
if ($argc < 2) {
	say STDERR "Usage: $0 <destination[:offset]> <<source|->[:offset]> [len]";
	say STDERR "`note: source is '-' means that 'fd_in' equal 'fd_out' # see copy_file_range(2)";
	say STDERR "  e.g1: $0 testfile:256  -  64";
	say STDERR "  e.g2: $0 fileout  filein  1024";
	exit 1;
}

#parse args
my ($dst_file, $off_out) = split(/:/, $ARGV[0]);
my ($src_file, $off_in) = split(/:/, $ARGV[1]);

#open test files
my $fd_in = 0;
my $fd_out = 0;
$fd_out = POSIX::open($dst_file, &POSIX::O_RDWR|&POSIX::O_CREAT, 0666) ||
	die "POSIX::open $dst_file fail: $!";
if ($src_file eq "-" || $src_file eq "") {
	$fd_in = $fd_out;
} else {
	$fd_in = POSIX::open($src_file, &POSIX::O_RDONLY) ||
		die "POSIX::open $src_file fail: $!";
}

#get offset and len
my $len = -s $src_file;
if ($fd_in == $fd_out && $off_out eq "") {
	$off_out = $len;
}
$off_out = 0 if ($off_out eq "");
$off_in = 0 if ($off_in eq "");

$off_in = pack("Q",$off_in);
$off_out = pack("Q",$off_out);
$len = 0+sprintf("%zu", $ARGV[2]) if ($argc >= 3);  #must convert to Integer

say STDOUT "[debug]: fd_in=$fd_in fd_out=$fd_out off_in=$off_in off_out=$off_out len=$len";
#copy_file_range
my $ret = 0;
while (1) {
	$ret = syscall(SYS_copy_file_range(),
			$fd_in, $off_in,
			$fd_out, $off_out,
			$len, 0);
	die "SYS_copy_file_range fail: $!" if ($ret == -1);

	say STDOUT "ret=$ret, len=$len";
	$len -= $ret;
	last if ($len <= 0);
	last if (eof($fd_in));
}

close($fd_in);
close($fd_out) if ($fd_in != $fd_out);
